[dcl.dcl] Replace 'could' and 'might' (#4362)
cplusplus/draft: source/declarations.tex

For a constexpr function or constexpr constructor
that is neither defaulted nor a template,
if no argument values exist such that

> an invocation of the function or constructor would be an evaluated subexpression of a core

Reviewer: I find this very hard to understand with "would" here. I think part of the problem is that the original wording is unclear. What we mean is:

"For a constexpr function or constexpr constructor that is neither defaulted nor a template, if there does not exist a valid core constant expression whose evaluation would evaluate an invocation of the function or constructor, and, for a constructor, there does not exist a valid declaration of a constant-initialized variable whose evaluation would evaluate an invocation of the function or constructor."

That is, this is really "for all possible core constant expressions", not "for all possible argument values".

I think we can simplify and clarify that:

"A constexpr function that is neither defaulted nor a template is ill-formed, no diagnostic required, if it is not possible for an evaluation of an invocation of the function to be performed while evaluating any valid manifestly constant-evaluated expression."

Author: fixed

a redundant set of parentheses around a parameter name and an object declaration
with a function-style cast as the initializer.
Just as for the ambiguities mentioned in~\ref{stmt.ambig},

> the resolution is to consider any construct that can possibly
> be a declaration a declaration.

Reviewer: ```suggestion
the resolution is to consider any construct
that matches the syntax of a declaration
to be a declaration.
```

An ambiguity can arise from the similarity between a function-style
cast and a
\grammarterm{type-id}.

> The resolution is that any construct that can possibly be a
> \grammarterm{type-id}
> in its syntactic context shall be considered a
> \grammarterm{type-id}.

Reviewer: ```suggestion
The resolution is that any construct
that matches the syntax of a \grammarterm{type-id}
is interpreted as a \grammarterm{type-id}.
```

no constructor of the object's class
is invoked for the initialization.
\begin{note}
Such an object can use value-initializion,

> or aggregate initialization\iref{dcl.init.aggr}

Reviewer: ```suggestion
aggregate initialization\iref{dcl.init.aggr},
```

int x = 999;                    // \tcode{x} is not a constant expression
const int y = 999;
const int z = 99;
char c1 = x;                    // OK, though it can narrow (in this case, it does narrow)

> char c2{x};                     // error: can narrow

Reviewer: I don't think "can" makes sense here, because the code is ill-formed, so it can't do anything. How about "error: potentially narrows"?

Author: fixed

signed int si1 =
  { (unsigned int)-1 };         // error: narrows
int ii = {2.0};                 // error: narrows

> float f1 { x };                 // error: can narrow

Reviewer: Likewise.

Author: fixed

Function \tcode{g}'s second parameter has a \tcode{carries_dependency} attribute,
but its first parameter does not. Therefore, function \tcode{h}'s first call to
\tcode{g} carries a dependency into \tcode{g}, but its second call does not. The

> implementation can be required to insert a fence prior to the second call to

Reviewer: "can be required" sounds like the programmer can cause the implementation to be required to do this, which isn't really right. This solves that problem but I'm still not overjoyed by it:
```suggestion
implementation can need to insert a fence prior to the second call to
```

Author: "can need"?

ugh

"It is possible that the implementation needs to insert a fence..."

\pnum
\recommended
The use of a fallthrough statement should suppress

> a warning that an implementation would otherwise issue

Reviewer: ```suggestion
any warning that an implementation would otherwise issue
```
(To avoid the suggestion that there would necessarily be such a warning.)
